python/ray/util/spark/cluster_init.py

```diff
@@ -797,16 +776,148 @@ def init_ray_cluster(
             "by `ray.shutdown()` before initiating a Ray cluster on spark."
         )
 
-    cluster = _init_ray_cluster(
+    spark = get_spark_session()
+
+    spark_master = spark.sparkContext.master
+    if not (
+        spark_master.startswith("spark://") or spark_master.startswith("local-cluster[")
+    ):
+        raise RuntimeError(
+            "Ray on Spark only supports spark cluster in standalone mode or local-cluster mode"
+        )
+
+    if (
+        is_in_databricks_runtime() and
+        Version(os.environ["DATABRICKS_RUNTIME_VERSION"]).major >= 12
+    ):
+        support_stage_scheduling = True
+    elif Version(pyspark.__version__).release >= (3, 4, 0):
+        support_stage_scheduling = True
+    else:
+        support_stage_scheduling = False
+
+    # Environment configurations within the Spark Session that dictate how many cpus
+    # and gpus to use for each submitted spark task.
+    num_spark_task_cpus = int(spark.sparkContext.getConf().get("spark.task.cpus", "1"))
+
+    if num_cpus_per_node is not None and num_cpus_per_node <= 0:
+        raise ValueError("Argument `num_cpus_per_node` value must be > 0.")
+
+    num_spark_task_gpus = int(
+        spark.sparkContext.getConf().get("spark.task.resource.gpu.amount", "0")
+    )
+
+    if num_gpus_per_node is not None and num_spark_task_gpus == 0:
+        raise ValueError(
+            "The spark cluster is not configured with 'gpu' resources, so that "
+            "you cannot specify the `num_gpus_per_node` argument."
+        )
+
+    if num_gpus_per_node is not None and num_gpus_per_node < 0:
+        raise ValueError("Argument `num_gpus_per_node` value must be >= 0.")
+
+    if num_cpus_per_node is not None or num_gpus_per_node is not None:
+        if support_stage_scheduling:
+            num_cpus_per_node = num_cpus_per_node or num_spark_task_cpus
+            num_gpus_per_node = num_gpus_per_node or num_spark_task_gpus
+
+            using_stage_scheduling = True
+            res_profile = _create_resource_profile(
+                num_cpus_per_node, num_gpus_per_node
+            )
+        else:
+            raise ValueError(
+                "Current spark version does not support stage scheduling, so that "
+                "you cannot set the argument `num_cpus_per_node` and "
+                "`num_gpus_per_node` values. Without setting the 2 arguments, "
+                "per-Ray worker node will be assigned with number of "
+                f"'spark.task.cpus' (equals to {num_spark_task_cpus}) cpu cores "
+                "and number of 'spark.task.resource.gpu.amount' "
+                f"(equals to {num_spark_task_gpus}) GPUs. To enable spark stage "
+                "scheduling, you need to upgrade spark to 3.4 version or use "
+                "Databricks Runtime 12.x."
```

Comment:
Databricks 12.0 uses Spark 3.3. Stage level scheduling is still supported in that case?

Reply:
Yes because databricks cherry-picks the feature to 12.0